server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java

```diff
@@ -91,63 +116,226 @@ private void registerListeners() {
+    // health node reported in the previous cluster state to be safe that we do not miss any change due to
+    // a flaky state.
+    private boolean hasHealthNodeChanged(DiscoveryNode currentHealthNode, ClusterChangedEvent event) {
+        DiscoveryNode previousHealthNode = HealthNode.findHealthNode(event.previousState());
+        return Objects.equals(lastSeenHealthNode.get(), currentHealthNode == null ? null : currentHealthNode.getId()) == false
+            || Objects.equals(previousHealthNode, currentHealthNode) == false;
     }
 
+    @Nullable
     DiskHealthInfo getLastReportedDiskHealthInfo() {
-        return lastReportedDiskHealthInfo;
+        return lastReportedDiskHealthInfo.get();
+    }
+
+    /**
+     * This class is responsible for repeatedly running the health monitoring. The first execution happens
+     * upon initialization and the following at the predefined interval. If there is an exception,
+     * it will log it and continue to schedule the next execution unless it was a EsRejectedExecutionException
+     */
+    static class HealthMonitoringTask implements Runnable, Scheduler.Cancellable {
+
+        private final TimeValue interval;
+        private final String executor;
+        private final Scheduler scheduler;
+        private final ClusterService clusterService;
+        private final DiskCheck diskCheck;
+        private final Client client;
+
+        private volatile boolean run = true;
+        private final AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo;
+        private final AtomicReference<String> lastSeenHealthNode;
+
+        HealthMonitoringTask(
+            TimeValue interval,
+            String executor,
+            Scheduler scheduler,
+            AtomicReference<DiskHealthInfo> lastReportedDiskHealthInfo,
+            AtomicReference<String> lastSeenHealthNode,
+            DiskCheck diskCheck,
+            ClusterService clusterService,
+            Client client
+        ) {
+            this.interval = interval;
+            this.executor = executor;
+            this.scheduler = scheduler;
+            this.lastReportedDiskHealthInfo = lastReportedDiskHealthInfo;
+            this.lastSeenHealthNode = lastSeenHealthNode;
+            this.clusterService = clusterService;
+            this.diskCheck = diskCheck;
+            this.client = client;
+            scheduler.schedule(this, TimeValue.ZERO, executor);
+        }
+
+        @Override
+        public boolean cancel() {
+            final boolean result = run;
+            run = false;
+            return result;
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return run == false;
+        }
+
+        /**
+         * This method evaluates the health info of this node and if there is a change it sends an update request to the health node
+         */
+        @Override
+        public void run() {
+            boolean nextRunScheduled = false;
```

Comment:
We probably ought to check the value of the `run` variable here right? Since it could be that this was scheduled to run in 30 seconds in the future in `scheduleNextRunIfNecessary()` and was cancelled at some point in that 30 seconds.

Reply:
Yes indeed. Good catch!